go/arrow/compute/internal/kernels/vector_run_end_encode.go

```diff
@@ -0,0 +1,947 @@
+	valueType                  arrow.DataType
+
+	estimatedValuesLen int64
+}
+
+func (re *runEndEncodeLoopBinary[R, O]) readValue(idx int64) ([]byte, bool) {
+	if len(re.inputValidity) > 0 && bitutil.BitIsNotSet(re.inputValidity, int(idx)) {
+		return nil, false
+	}
+
+	start, end := re.offsetValues[idx], re.offsetValues[idx+1]
+	return re.inputValues[start:end], true
+}
+
+func (re *runEndEncodeLoopBinary[R, O]) CountNumberOfRuns() (numValid, numOutput int64) {
+	re.estimatedValuesLen = 0
+
+	var offset int64
+	currentRun, curRunValid := re.readValue(offset)
+	offset++
+
+	if curRunValid {
+		numValid = 1
+		re.estimatedValuesLen += int64(len(currentRun))
+	}
+	numOutput = 1
+
+	for offset < re.inputLen {
+		value, valid := re.readValue(offset)
+		offset++
+		// new run
+		if valid != curRunValid || !bytes.Equal(value, currentRun) {
+			if valid {
+				re.estimatedValuesLen += int64(len(value))
+			}
+
+			currentRun = value
+			curRunValid = valid
+
+			numOutput++
+			if valid {
+				numValid++
+			}
+		}
+	}
+	return
+}
+
+func (re *runEndEncodeLoopBinary[R, O]) PreallocOutput(ctx *exec.KernelCtx, numOutput int64, out *exec.ExecResult) {
+	runEndsBuffer := ctx.Allocate(int(numOutput) * int(SizeOf[R]()))
+	var validityBuffer *memory.Buffer
+	if len(re.inputValidity) > 0 {
+		validityBuffer = ctx.AllocateBitmap(numOutput)
+	}
+
+	valueBuffer := ctx.Allocate(int(re.estimatedValuesLen))
+	offsetsBuffer := ctx.Allocate(int(numOutput+1) * int(SizeOf[O]()))
+
+	reeType := arrow.RunEndEncodedOf(exec.GetDataType[R](), re.valueType)
+	*out = exec.ExecResult{
+		Type:   reeType,
+		Len:    re.inputLen,
+		Nulls:  0,
+		Offset: 0,
+		Children: []exec.ArraySpan{
+			{
+				Type: reeType.RunEnds(),
+				Len:  numOutput,
+			},
+			{
+				Type: reeType.Encoded(),
+				Len:  numOutput,
+			},
+		},
+	}
+
+	out.Children[0].Buffers[1].WrapBuffer(runEndsBuffer)
+	if validityBuffer != nil {
+		out.Children[1].Buffers[0].WrapBuffer(validityBuffer)
+	}
+	out.Children[1].Buffers[1].WrapBuffer(offsetsBuffer)
+	out.Children[1].Buffers[2].WrapBuffer(valueBuffer)
+}
+
+func (re *runEndEncodeLoopBinary[R, O]) WriteEncodedRuns(out *exec.ExecResult) int64 {
+	outputRunEnds := exec.GetData[R](out.Children[0].Buffers[1].Buf)
+	outputOffsets := exec.GetSpanOffsets[O](&out.Children[1], 1)
+	outputValues := out.Children[1].Buffers[2].Buf
+
+	var readOffset int64
+	currentRun, curRunValid := re.readValue(readOffset)
+	readOffset++
+
+	var writeOffset, valueOffset int64
+	validityBuf := out.Children[1].Buffers[0].Buf
+	setValidity := func(valid bool) {}
+	if len(validityBuf) > 0 {
+		setValidity = func(valid bool) {
+			bitutil.SetBitTo(validityBuf, int(writeOffset), valid)
+		}
+	}
+
+	outputOffsets[0], outputOffsets = 0, outputOffsets[1:]
+
+	writeValue := func(valid bool, value []byte) {
+		setValidity(valid)
+		valueOffset += int64(copy(outputValues[valueOffset:], value))
+		outputOffsets[writeOffset] = O(valueOffset)
+	}
+
+	for readOffset < re.inputLen {
```

Comment:
`readOffset < re.inputOffset + re.inputLen`

Either that or I'm missing something about Binary array types.